source/common/upstream/upstream_impl.cc

```diff
@@ -87,6 +87,65 @@ HostsPerLocalityImpl::filter(std::function<bool(const Host&)> predicate) const {
   return shared_filtered_clone;
 }
 
+void HostSetImpl::updateHosts(HostVectorConstSharedPtr hosts,
+                              HostVectorConstSharedPtr healthy_hosts,
+                              HostsPerLocalityConstSharedPtr hosts_per_locality,
+                              HostsPerLocalityConstSharedPtr healthy_hosts_per_locality,
+                              LocalityWeightsConstSharedPtr locality_weights,
+                              const HostVector& hosts_added, const HostVector& hosts_removed) {
+  hosts_ = std::move(hosts);
+  healthy_hosts_ = std::move(healthy_hosts);
+  hosts_per_locality_ = std::move(hosts_per_locality);
+  healthy_hosts_per_locality_ = std::move(healthy_hosts_per_locality);
+  locality_weights_ = std::move(locality_weights);
+  // Rebuild the locality scheduler.
+  if (hosts_per_locality_ != nullptr && locality_weights_ != nullptr &&
+      !locality_weights_->empty()) {
+    locality_scheduler_ = std::make_unique<EdfScheduler<LocalityEntry>>();
```

Comment:
I agree; as mentioned above, we should probably lazily rebuild as we age entries out of the scheduler during picks. I think it's better as a TODO to tune as this starts to matter (rebuilding with few localities is still really cheap). We would need to first verify that the locality indexes haven't changed (i.e. the map from host vector per locality to absolute locality), then we could just update the local map; this might cost more than just rebuilding the scheduler for a small number of localities.